src/downloader/downloader.cpp

```diff
@@ -0,0 +1,134 @@
+
+Downloader::~Downloader()
+{
+    std::lock_guard<std::mutex> lock {mMutex};
+
+    mShutdown = true;
+    mCondVar.notify_all();
+}
+
+Error Downloader::Download(const String& url, const String& path, aos::downloader::DownloadContent contentType)
+{
+    LOG_DBG() << "Sync downloading: url=" << url << ", path=" << path << ", contentType=" << contentType;
+
+    return RetryDownload(url, path);
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+Error Downloader::Download(const String& url, const String& path)
+{
+    Poco::URI uri(url.CStr());
+    if (uri.getScheme() == "file") {
+        return CopyFile(uri, path);
+    }
+
+    std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> curl(curl_easy_init(), curl_easy_cleanup);
+    if (!curl) {
+        return Error(ErrorEnum::eFailed, "Failed to init curl");
+    }
+
+    auto fileCloser = [](FILE* fp) {
+        if (fp) {
+            if (auto res = fclose(fp); res != 0) {
+                LOG_ERR() << "Failed to close file: res=" << res;
+            }
+        }
+    };
+
+    std::unique_ptr<FILE, decltype(fileCloser)> fp(fopen(path.CStr(), "ab"), fileCloser);
+    if (!fp) {
+        return Error(ErrorEnum::eFailed, "Failed to open file");
+    }
+
+    fseek(fp.get(), 0, SEEK_END);
+
+    auto existingFileSize = ftell(fp.get());
+
+    curl_easy_setopt(curl.get(), CURLOPT_URL, url.CStr());
+    curl_easy_setopt(curl.get(), CURLOPT_RESUME_FROM_LARGE, existingFileSize);
+    curl_easy_setopt(curl.get(), CURLOPT_WRITEFUNCTION, fwrite);
+    curl_easy_setopt(curl.get(), CURLOPT_WRITEDATA, fp.get());
+    curl_easy_setopt(curl.get(), CURLOPT_TIMEOUT, cTimeoutSec); // Timeout in seconds
+    curl_easy_setopt(curl.get(), CURLOPT_CONNECTTIMEOUT, cTimeoutSec);
+
+    auto res = curl_easy_perform(curl.get());
+    if (res != CURLE_OK) {
+        return Error(ErrorEnum::eFailed, curl_easy_strerror(res));
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Downloader::CopyFile(const Poco::URI& uri, const String& outfilename)
+{
+    auto path = uri.getPath();
+    if (path.empty() && !uri.getHost().empty()) {
+        path = uri.getHost();
+    }
+
+    if (!std::filesystem::exists(path)) {
+        return Error(ErrorEnum::eFailed, "File not found");
+    }
+
+    try {
+        std::ifstream ifs(path, std::ios::binary);
+        std::ofstream ofs(outfilename.CStr(), std::ios::binary | std::ios::trunc);
+
+        ofs << ifs.rdbuf();
+
+        return ErrorEnum::eNone;
+    } catch (const std::exception& e) {
+        return Error(ErrorEnum::eFailed, e.what());
+    }
+}
+
+Error Downloader::RetryDownload(const String& url, const String& path)
+{
+    auto  delay = cDelay;
+    Error error;
```

Comment:
Looks like `err` is a pattern naming for error object
```suggestion
    Error err;
```

Reply:
Done